File: frontend/src/lib/hosts.ts
/**
 * Shared TypeScript types and small helpers for the host dashboard.
 *
 * The fields mirror the backend Pydantic schemas (see
 * backend/vman/schemas/hosts.py) but the wire shape is kept narrow on
 * purpose: we never accept nor display any plaintext credential
 * material. `credential_id` is an opaque reference to an entry in the
 * encrypted vault; the form lets the user create / pick one but never
 * shows the secret itself.
 */

export type AuthMethod = "password" | "key" | "key_with_passphrase";
export type SudoMode = "root" | "passwordless_sudo" | "sudo_password";
export type Environment = "experiment" | "staging" | "production";
export type RiskLevel = "low" | "medium" | "high" | "critical" | null;

export interface Host {
  id: string;
  name: string;
  hostname_or_ip: string;
  ssh_port: number;
  username: string;
  auth_method: AuthMethod;
  credential_id: string | null;
  sudo_mode: SudoMode;
  host_key_fingerprint: string | null;
  host_key_algorithm: string | null;
  os_family: string | null;
  os_name: string | null;
  os_version: string | null;
  package_manager: string | null;
  arch: string | null;
  cpu_cores: number | null;
  ram_mb: number | null;
  disk_total_mb: number | null;
  provider: string | null;
  region: string | null;
  environment: Environment;
  risk_level: RiskLevel;
  tags: string[];
  notes: string;
  last_seen_at: string | null;
  disabled_at: string | null;
  created_at: string;
  updated_at: string;
}

/**
 * The shape the dashboard sends when creating a host. We deliberately
 * leave secret material out: `credential_id` references a vault entry
 * that the user must provision separately.
 */
export interface HostCreatePayload {
  name: string;
  hostname_or_ip: string;
  ssh_port: number;
  username: string;
  auth_method: AuthMethod;
  credential_id?: string | null;
  sudo_mode: SudoMode;
  environment: Environment;
  provider?: string | null;
  region?: string | null;
  tags: string[];
  notes: string;
}

export type HostUpdatePayload = Partial<HostCreatePayload> & {
  risk_level?: RiskLevel;
  host_key_fingerprint?: string | null;
  host_key_algorithm?: string | null;
  os_family?: string | null;
  os_name?: string | null;
  os_version?: string | null;
  package_manager?: string | null;
  arch?: string | null;
  cpu_cores?: number | null;
  ram_mb?: number | null;
  disk_total_mb?: number | null;
};

export const AUTH_METHODS: { value: AuthMethod; label: string; hint: string }[] = [
  { value: "key", label: "SSH key (recommended)", hint: "Authenticates with a public key on the target." },
  { value: "key_with_passphrase", label: "SSH key + passphrase", hint: "Key is encrypted with an additional passphrase." },
  { value: "password", label: "Password", hint: "Use only when keys are not possible. Stored encrypted." },
];

export const SUDO_MODES: { value: SudoMode; label: string; hint: string }[] = [
  { value: "root", label: "Root login", hint: "The SSH user is root. No sudo needed." },
  { value: "passwordless_sudo", label: "Passwordless sudo", hint: "User can run any command via sudo NOPASSWD." },
  { value: "sudo_password", label: "Sudo with password", hint: "User must enter a password for sudo." },
];

export const ENVIRONMENTS: { value: Environment; label: string; tone: "info" | "warning" | "destructive" }[] = [
  { value: "experiment", label: "Experiment", tone: "info" },
  { value: "staging", label: "Staging", tone: "warning" },
  { value: "production", label: "Production", tone: "destructive" },
];

export function environmentLabel(env: Environment): string {
  return ENVIRONMENTS.find((e) => e.value === env)?.label ?? env;
}

export function authMethodLabel(method: AuthMethod): string {
  return AUTH_METHODS.find((m) => m.value === method)?.label ?? method;
}

export function sudoModeLabel(mode: SudoMode): string {
  return SUDO_MODES.find((m) => m.value === mode)?.label ?? mode;
}

/**
 * Best-effort short tag for the OS column. We never claim more than
 * the server told us, so missing data renders as a dash.
 */
export function describeOs(host: Host): string {
  const parts: string[] = [];
  if (host.os_name) parts.push(host.os_name);
  if (host.os_version) parts.push(host.os_version);
  if (host.arch) parts.push(host.arch);
  return parts.length > 0 ? parts.join(" • ") : "—";
}

/**
 * The connection-test result shape returned by
 * ``POST /api/hosts/{id}/test`` (or the corresponding mock when the
 * backend endpoint is not yet wired up).
 */
export interface ConnectionTestResult {
  ok: boolean;
  reached: boolean;
  authenticated: boolean;
  host_key_fingerprint: string | null;
  host_key_algorithm: string | null;
  latency_ms: number | null;
  message: string;
  tested_at: string;
}

export const HOST_DETAIL_FIELD_LABELS: Record<string, string> = {
  name: "Name",
  hostname_or_ip: "Hostname / IP",
  ssh_port: "SSH port",
  username: "SSH user",
  auth_method: "Auth method",
  sudo_mode: "Sudo mode",
  environment: "Environment",
  provider: "Provider",
  region: "Region",
  tags: "Tags",
  notes: "Notes",
  host_key_fingerprint: "Host key fingerprint",
  host_key_algorithm: "Host key algorithm",
  os_family: "OS family",
  os_name: "OS name",
  os_version: "OS version",
  package_manager: "Package manager",
  arch: "Architecture",
  cpu_cores: "CPU cores",
  ram_mb: "RAM",
  disk_total_mb: "Disk",
  risk_level: "Risk level",
  credential_id: "Credential",
  last_seen_at: "Last seen",
  disabled_at: "Disabled at",
  created_at: "Created at",
  updated_at: "Updated at",
};

File: backend/vman/schemas/hosts.py
"""Pydantic schemas for the host CRUD API."""

from __future__ import annotations

import ipaddress
import re

from pydantic import BaseModel, Field, field_validator

_NAME_RE = re.compile(r"^[a-zA-Z0-9][a-zA-Z0-9_.-]{0,63}$")


def _validate_ip_or_host(value: str) -> str:
    """Accept either a literal IPv4/IPv6 address OR a DNS hostname.

    DNS hostnames are not strictly validated (DNS allows many
    characters); we just reject empty / overly long values and
    obvious garbage.
    """
    if not value or len(value) > 255:
        raise ValueError("hostname_or_ip must be 1-255 chars")
    # First try parsing as an IP address.
    try:
        ipaddress.ip_address(value)
        return value
    except ValueError:
        pass
    # Otherwise treat as a DNS hostname. Very loose validation.
    if any(c.isspace() for c in value):
        raise ValueError("hostname_or_ip must not contain whitespace")
    return value


class HostCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=64)
    hostname_or_ip: str = Field(..., min_length=1, max_length=255)
    ssh_port: int = Field(22, ge=1, le=65535)
    username: str = Field(..., min_length=1, max_length=64)
    auth_method: str = Field(..., min_length=1, max_length=32)
    credential_id: str | None = Field(default=None, max_length=64)
    sudo_mode: str = Field("root", max_length=32)
    environment: str = Field("experiment", max_length=16)
    provider: str | None = Field(default=None, max_length=64)
    region: str | None = Field(default=None, max_length=64)
    tags: list[str] = Field(default_factory=list, max_length=64)
    notes: str = Field("", max_length=2048)

    @field_validator("name")
    @classmethod
    def _validate_name(cls, v: str) -> str:
        if not _NAME_RE.match(v):
            raise ValueError("name must start with alnum, then [a-zA-Z0-9_.-], max 64 chars")
        return v

    @field_validator("hostname_or_ip")
    @classmethod
    def _validate_host(cls, v: str) -> str:
        return _validate_ip_or_host(v)

    @field_validator("auth_method")
    @classmethod
    def _validate_auth_method(cls, v: str) -> str:
        if v not in {"password", "key", "key_with_passphrase"}:
            raise ValueError("auth_method must be 'password', 'key', or 'key_with_passphrase'")
        return v

    @field_validator("sudo_mode")
    @classmethod
    def _validate_sudo_mode(cls, v: str) -> str:
        if v not in {"root", "passwordless_sudo", "sudo_password"}:
            raise ValueError("sudo_mode must be 'root', 'passwordless_sudo', or 'sudo_password'")
        return v

    @field_validator("environment")
    @classmethod
    def _validate_environment(cls, v: str) -> str:
        if v not in {"experiment", "staging", "production"}:
            raise ValueError("environment must be 'experiment', 'staging', or 'production'")
        return v


class HostUpdate(BaseModel):
    """Partial update payload; every field is optional."""

    hostname_or_ip: str | None = Field(default=None, max_length=255)
    ssh_port: int | None = Field(default=None, ge=1, le=65535)
    username: str | None = Field(default=None, min_length=1, max_length=64)
    auth_method: str | None = Field(default=None, min_length=1, max_length=32)
    credential_id: str | None = Field(default=None, max_length=64)
    sudo_mode: str | None = Field(default=None, max_length=32)
    environment: str | None = Field(default=None, max_length=16)
    provider: str | None = Field(default=None, max_length=64)
    region: str | None = Field(default=None, max_length=64)
    tags: list[str] | None = Field(default=None, max_length=64)
    notes: str | None = Field(default=None, max_length=2048)
    host_key_fingerprint: str | None = Field(default=None, max_length=128)
    host_key_algorithm: str | None = Field(default=None, max_length=16)
    os_family: str | None = Field(default=None, max_length=32)
    os_name: str | None = Field(default=None, max_length=64)
    os_version: str | None = Field(default=None, max_length=64)
    package_manager: str | None = Field(default=None, max_length=32)
    arch: str | None = Field(default=None, max_length=16)
    cpu_cores: int | None = Field(default=None, ge=0)
    ram_mb: int | None = Field(default=None, ge=0)
    disk_total_mb: int | None = Field(default=None, ge=0)
    risk_level: str | None = Field(default=None, max_length=16)

    @field_validator("hostname_or_ip")
    @classmethod
    def _validate_host(cls, v: str | None) -> str | None:
        if v is None:
            return v
        return _validate_ip_or_host(v)

    @field_validator("auth_method")
    @classmethod
    def _validate_auth_method(cls, v: str | None) -> str | None:
        if v is None:
            return v
        if v not in {"password", "key", "key_with_passphrase"}:
            raise ValueError("auth_method must be 'password', 'key', or 'key_with_passphrase'")
        return v

    @field_validator("sudo_mode")
    @classmethod
    def _validate_sudo_mode(cls, v: str | None) -> str | None:
        if v is None:
            return v
        if v not in {"root", "passwordless_sudo", "sudo_password"}:
            raise ValueError("sudo_mode must be 'root', 'passwordless_sudo', or 'sudo_password'")
        return v

    @field_validator("environment")
    @classmethod
    def _validate_environment(cls, v: str | None) -> str | None:
        if v is None:
            return v
        if v not in {"experiment", "staging", "production"}:
            raise ValueError("environment must be 'experiment', 'staging', or 'production'")
        return v


class HostOut(BaseModel):
    id: str
    name: str
    hostname_or_ip: str
    ssh_port: int
    username: str
    auth_method: str
    credential_id: str | None = None
    sudo_mode: str
    host_key_fingerprint: str | None = None
    host_key_algorithm: str | None = None
    os_family: str | None = None
    os_name: str | None = None
    os_version: str | None = None
    package_manager: str | None = None
    arch: str | None = None
    cpu_cores: int | None = None
    ram_mb: int | None = None
    disk_total_mb: int | None = None
    provider: str | None = None
    region: str | None = None
    environment: str
    risk_level: str | None = None
    tags: list[str]
    notes: str
    last_seen_at: str | None = None
    disabled_at: str | None = None
    created_at: str
    updated_at: str


class ConnectionTestResult(BaseModel):
    ok: bool
    reached: bool
    authenticated: bool
    host_key_fingerprint: str | None = None
    host_key_algorithm: str | None = None
    latency_ms: float | None = None
    message: str
    tested_at: str
    # Detected fields so the dashboard can refresh without a second GET.
    os_family: str | None = None
    os_name: str | None = None
    os_version: str | None = None
    package_manager: str | None = None
    arch: str | None = None
    cpu_cores: int | None = None
    ram_mb: int | None = None
    disk_total_mb: int | None = None
    last_seen_at: str | None = None


__all__ = ["HostCreate", "HostOut", "HostUpdate", "ConnectionTestResult"]

